### ISSUE TITLE
Notebooks Formatting/Style

### ISSUE THREAD

**Issue by MrDupin** [aimacode/aima-python#526]
I believe the Notebooks would benefit a lot if we adhered to a clear(-er) format. Below I list some of the things that stand out.

## The 'Nitpicks'

The issues here are small and can be easily fixed, but I think they are worth mentioning even though they are low priority edits.

a) Styling: A lot of times we alternate the styling in our text. In one section we can have one styling for code and in the next another (`function_name` in one, *function_name* in another).

b) Headings/Titles: This is particularly evident in the [agents notebook](https://github.com/aimacode/aima-python/blob/master/agents.ipynb). Usually we have titles lowercase, but here we alternate. Also, the size of the heading (determined by the number of # we use) is not consistent across notebooks.

c) Contents: Most of the notebooks miss contents. I think they are very important for the reader. If someone wants to read on a certain algorithm, they should be able to quickly glance at a notebook and see if the algorithm is in.

## The More Serious Difference

Usually when explaining an algorithm, we first write a quick overview, then go into implementation details and finally we solve an example (or multiple examples). Other times we do all of them together. We run an example and explain as we go along.

An example for the first technique is the Plurality Learner in the [learning module](https://github.com/aimacode/aima-python/blob/master/learning.ipynb). Even though it is a simple and short algorithm, we can see that first we explain the algorithm, then the pseudocode and then run an example (the sections in the example are titled accordingly, eg. 'Overview', but that is not necessary). An example for the other technique is the Backtracking Search in the [csp module](https://github.com/aimacode/aima-python/blob/master/csp.ipynb). We write 1-2 sentences explaining what the algorithm does, then go a little bit in the implementation before jumping into an example. As we run the example we explain how the implementation and the algorithm work.

While there are merits to both techniques, I believe it would be best to pick one for consistency's sake. When a student reads a module, it would be easier on them if the rest of the modules followed the same format. When a student expects an explanation of the algorithm and instead we jump into an example, it might be hard to follow initially.

I personally believe the first technique is better, but it is open to discussion.

---

These are the main issues with format/style consistency in the Notebooks. I believe it is something we should be addressing over time. I would like to discuss this now before more and more Notebooks are added and becomes difficult to fix.

**Comment by kaivalyar** [comment]
I agree completely with your analysis of the current state of the notebooks, however, I don't see the need to standardize the approach we use (this is not with reference the "nitpicks" - those should definitely be made standard to improve clarity). I think it's okay to sometimes intersperse examples when explaining an algorithm, and sometimes to not do so. In my university, different approaches are used by different professors, and both have been successful to various degrees. Maybe it's better to evaluate this on a case by case basis?

**Comment by MrDupin** [comment]
In different professors you can see different approaches because of the professor's preferences and the subject, which is different itself. Students don't expect such consistency.

In the same book though, you rarely see such variations without an important reason. What we are doing is closer to a book, so it would benefit if students knew what they were getting into on each algorithm.

**Comment by norvig** [comment]
I would suggest:

* Let's always use `function_name`, not *function_name*. Code is code.
* Section Headers Should Have Capitalized Words
* A Table of Contents is very useful, especially if they have links; Is there an easy way to automatically generate it?
* I agree that the "first technique" is usually better, but sometimes concepts are complex, and it is ok to mix approaches, doing a little of one thing, then returning to do more. So I don't think we should require consistency here; rather we should focus on clarity -- if it is clear, the technique used is less important.
